### ISSUE TITLE
Regression: tab_size settings being ignored

### ISSUE THREAD

**Issue by fserb** [zed-industries/zed#12154]
### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

I have `"tab_size": 2` on my settings. With the latest Zed stable (0.136.2) and preview (0.137.0), when I'm editing most file types (Shell Script, Plain Text, Go, HTML, Python) and press tab, I get moved 4 characters ahead. This also happens on auto-indent and auto-format.

This doesn't seem to happen (and I get the correct 2 characters) on Javascript, YAML, JSON files.

This wasn't happening on the previous version.

### Environment

Zed: v0.136.2 (Zed)
OS: macOS 14.5.0
Memory: 64 GiB
Architecture: aarch64

**Comment by maxdeviant** [comment]
@fserb Are you able to share the contents of your `settings.json`?

**Comment by fserb** [comment]
Sure, give me a sec. 

But I've also noticed that other settings are being ignored. And the problem seems to only happen if I've a `"languages"` section on my settings. If I comment it out, everything works fine. Even a `languages: {},` seems to cause all sorts of weird behavior (like auto-closing brackets coming back, etc).

**Comment by fserb** [comment]
I was able to reduce this to:

```
// Zed settings
{
  "tab_size": 2,
  "languages": {},
  "language_overrides": {}
}
```

This makes tab_size (and some other settings, but not themes, for example) to be completely ignored.

**Comment by fserb** [comment]
The reason I had both of them (from my actual config):

```
  "languages": {
    "JavaScript": {
      "formatter": {
        "external": {
          "command": "deno",
          "arguments": ["fmt", "-"]
        }
      }
    }
  },
  "language_overrides": {
    "JavaScript": {
      "language_servers": [
        "deno-language-server",
        "!typescript-language-server",
        "!vscode-language-server",
        "!eslint"
      ]
    }
},
```

Which I thought it was "the right way" of having Deno support.

**Comment by maxdeviant** [comment]
> The reason I had both of them (from my actual config):
> 
> ```
>   "languages": {
>     "JavaScript": {
>       "formatter": {
>         "external": {
>           "command": "deno",
>           "arguments": ["fmt", "-"]
>         }
>       }
>     }
>   },
>   "language_overrides": {
>     "JavaScript": {
>       "language_servers": [
>         "deno-language-server",
>         "!typescript-language-server",
>         "!vscode-language-server",
>         "!eslint"
>       ]
>     }
> },
> ```
> 
> Which I thought it was "the right way" of having Deno support.

`language_overrides` is an alias for `languages`, so you can't provide both of them.

Their contents can be merged together, however.

**Comment by JosephTLyons** [comment]
I just ran into this as well and was chatting with @bennetbo about it - glad to see some confirmation of this issue with others.

**Comment by emnul** [comment]
I'm also seeing this behavior when editing Rust files. Like @fserb said, this isn't affecting JSON files for me. Here's what I'm seeing on my end. Seems like some "off by one" calculation is occurring somewhere. My `tab_size` is set to 3 for context. As you can see, the spacing offset gets progressively worse by one additional space as the nesting level increases.

https://github.com/zed-industries/zed/assets/34749913/7f6231b3-7476-4a60-8b25-caa82039f466

**Comment by mike-ward** [comment]
I'm seeing a somewhat similar issue with the V language (Vlang), except tabs always display as 2.
```
{
  "format_on_save": "on",
  "formatter": {
    "external": {
      "command": "v",
      "arguments": ["fmt"]
    }
  },
  "autosave": { "after_delay": { "milliseconds": 500 } },
  "tab_size": 8,
  "hard_tabs": true
}
```
<img width="412" alt="image" src="https://github.com/zed-industries/zed/assets/567927/71f1e821-15df-4e26-aca5-7d3701c62f2d">

**Comment by mkozjak** [comment]
In my case with golang `"tab_size": 8` always shows one hard tab (4 whitespaces).

**Comment by bennetbo** [comment]
I think in some cases reported here zed actually works as expected. In the case of rust (I think the same goes for go and v) there is no way to tell `rustfmt` to use a different tab size other than 4. So if you have auto formatting enabled, zed will pick up your tab size correctly, however the formatter will still format with a fixed tab size because it is not configurable.
You can verify if zed is picking up the correct tab size, by pressing tab and see if the amount of inserted characters matches up with what you have configured.

Here's a showcase:

https://github.com/zed-industries/zed/assets/53836821/a16a8bf9-f21f-46b4-b86f-8abbc4aec29d

You can see that when I press tab, zed inserts 8 spaces (which is what I have configured), however when saving it goes back to tab size 4, which is what `rustfmt` enforces.

**Comment by mike-ward** [comment]
V formatter uses a tab size of 8.

**Comment by mkozjak** [comment]
> In my case with golang `"tab_size": 8` always shows one hard tab with width of 4 whitespaces, but when it saves the file and I check it with `cat` it's one hard tab with width of 8 whitespaces (how it's supposed to be in Buffer also, but is not).
> 
> What ultimately helped is adding this to my global config:
> 
> ```json
> "language_overrides": {
>     "Go": {
>         "tab_size": 8,
>         "hard_tabs": true
>     }
>   }
> ```

This doesn't help anymore. 🤔 
Version 0.142.6.

**Comment by SomeoneToIgnore** [comment]
`language_overrides` was deprecated in Zed 0.141.2

To quote https://zed.dev/releases/stable ,

```
SETTINGS DEPRECATION NOTICE

In your settings.json the following keys will need to be renamed:

    language_overrides -> languages
    copilot -> inline_completions
    show_copilot_suggestions -> show_inline_completions

Previously, we were supporting the old keys as an alias for the new keys, but this let to unintended/confusing failure modes. As a result we have deprecated the old keys.
```

**Comment by mkozjak** [comment]
> `language_overrides` was deprecated in Zed 0.141.2
> 
> To quote https://zed.dev/releases/stable ,
> 
> ```
> SETTINGS DEPRECATION NOTICE
> 
> In your settings.json the following keys will need to be renamed:
> 
>     language_overrides -> languages
>     copilot -> inline_completions
>     show_copilot_suggestions -> show_inline_completions
> 
> Previously, we were supporting the old keys as an alias for the new keys, but this let to unintended/confusing failure modes. As a result we have deprecated the old keys.
> ```

Thanks, that fixed it! 🎉

**Comment by DarkArc** [comment]
I'm seeing this in a way that kind of relates to https://github.com/zed-industries/zed/issues/14080

The tab_width is being applied for "what happens when I press the tab key" but it's not being applied for "how wide a tab should be rendered".

**Comment by Ekleog** [comment]
This is interesting, I have exactly the same issue as @DarkArc , but reversed (on 0.168.2).

With this config:
```json
{
  "tab_size": 4,
  "languages": {
    "Nix": {
      "tab_size": 2
    }
  },
  [...]
}
```

I'm seeing tab lines of 2 spaces but typing the Tab key moves to the nearest 4-spaces boundary.

But changing the global `tab_size` to 2, I have everything properly set at 2 spaces. So it looks like the `tab_size` setting is being half-set?

Also, if I now set both to 4, I do see 4 spaces per tab line and do get a 4-space-indenting tab key.

TL;DR: It looks to me like, on 0.168.2, the language-specific overrides for tab size only override the tab display size, and not the tab key alignment. However, the global tab_size setting does seem to set both properly.

**Comment by nt8r** [comment]
In at least one repository, I discovered that this setting is completely ignored if a `.editorconfig` file exists in the repository that specifies "indent_size" for the language. There's no indication in the UI that the setting is ignored in favor of the contents of the `.editorconfig` file, and #24108 seems to make the situation even worse.

For others experiencing these symptoms, check whether a `.editorconfig` file may be causing them.

**Comment by 0xtimsb** [comment]
After reading all the comments on this thread, I’ve figured there’s confusion between what `tab_size` does and the formatting of code, as also mentioned by @bennetbo.

1. `tab_size` simply determines how many characters it should add as spaces (or an actual hard tab, depending on your settings) after your cursor. After my testing, Zed seems to handle this correctly for all the languages mentioned in this thread. For this, I set `tab_size` to `3`, and it worked fine for all the example snippets I tested.

| Language      | Tab Indent on Key Press | Correct Tab Indent After Formatting      |
|---------------|-------------------------|------------------------------------------|
| JavaScript    | OK                      | OK                                       |
| JSON          | OK                      | OK                                       |
| YAML          | OK                      | OK                                       |
| Go            | OK                      | OK                                       |
| HTML          | OK                      | OK                                       |
| PHP           | OK                      | OK                                       |
| Python        | OK                      | OK (with ruff, upon changes)            |
| Rust          | OK                      | OK (with rustfmt, upon changes)         |
| Shell Script  | OK                      | OK (with shfmt)                         |
| C++           | OK                      | OK (with clang-format)                  |
| Nix           | OK                      | -                                        |
| V Lang        | OK                      | -                                        |

2. Upon saving, Zed uses `tab_size` to format code in some cases, while for others you have to manually configure this by passing an indent arguments to the formatter or by including a configuration file at the root of the project to explicitly specify it.

For example:  
- For Rust, in `rustfmt.toml`:  
```
 tab_spaces = 3 
```
- For Python, in `pyproject.toml`:
```
[tool.ruff]
indent-width = 3
``` 
- For C++, in `.clang-format`:  
```
BasedOnStyle: LLVM
IndentWidth: 3
TabWidth: 3
```
etc

Note: This is also how VS Code works.

3. Zed’s tab indentation is syntax-tree aware, which is a bit different from other editors.

https://github.com/user-attachments/assets/5cb4671f-9a80-4d56-ae41-258ce7ea17d6

For instance, when `tab_size` is set to `3` and your cursor is at an outdent position, it will indent to align with the expected syntax, potentially jumping `6` spaces in this case. However, for forward jumps, it falls back to `3`. This behavior is consistent across languages and intentional. I wonder if this is the tab size mismatch you’re all encountering.
 
---
If you still think there's some language-specific behavior I missed that isn't covered by the reasons mentioned above, feel free to reply.

**Comment by cody-quinn** [comment]
I'm having this issue on 174.6. Indentation is broken when tab_size is overridden for a specific language, when set globally it works as expected. See the attached video.

https://github.com/user-attachments/assets/86d62664-d5d8-498a-87d5-f82fcc402c71

**Comment by athre0z** [comment]
Same problem here, specifically when editing Nix files.

**Comment by smitbarmase** [comment]
@cody-quinn @athre0z can you provide your `settings.json`?

**Comment by athre0z** [comment]
## Config

```
// Zed settings
//
// For information on how to configure Zed, see the Zed
// documentation: https://zed.dev/docs/configuring-zed
//
// To see all of Zed's default settings without changing your
// custom settings, run the `open default settings` command
// from the command palette or from `Zed` application menu.
{
  "show_edit_predictions": true,
  "features": {
    "edit_prediction_provider": "zed"
  },
  "theme": {
    "mode": "system",
    "light": "One Light",
    "dark": "One Dark"
  },
  "wrap_guides": [80, 100],
  "format_on_save": "off",
  "remove_trailing_whitespace_on_save": false,
  "languages": {
    "C": {
      "show_edit_predictions": false,
      "show_inline_completions": true
    },
    "TypeScript": {
      "show_inline_completions": true
    },
    "Rust": {
      "show_edit_predictions": true,
      "show_inline_completions": false,
      "format_on_save": "on"
    },
    "JSON": { "format_on_save": "on" },
    "JSONC": { "format_on_save": "on" },
    "YAML": { "show_inline_completions": false, "format_on_save": "on" },
    "Go": { "show_inline_completions": false, "format_on_save": "on" },
    "Proto": { "format_on_save": "on" },
    "Nix": { "remove_trailing_whitespace_on_save": true, "tab_size": 2 }
  },
  "language_models": {
    "anthropic": {
      "version": "1"
    }
  },
  "autosave": "on_window_change",
  "assistant": {
    "enable_experimental_live_diffs": true,
    "default_model": {
      "provider": "anthropic",
      "model": "claude-3-7-sonnet-latest"
    },
    "version": "2",
    "provider": null
  },
  "show_inline_completions": false,
  "buffer_font_family": "JetBrains Mono",
  "terminal": {
    "font_family": "JetBrainsMono Nerd Font"
  }
}
```

## Recording

For some reason, the first tab inserts 2 characters, but all following one insert 4. Indenting/unindenting stuff by selecting + TAB / SHIFT+TAB always does 4 spaces.

https://github.com/user-attachments/assets/a5f783ce-986d-49e8-a498-460791a9492a

**Comment by cody-quinn** [comment]
```json5
{
  "vim_mode": true,
  "ui_font_size": 16,
  "buffer_font_size": 13,
  "buffer_font_family": "JetBrains Mono",
  "theme": {
    "mode": "dark",
    "dark": "Gruvbox Dark Hard",
    "light": "Atelier Dune Light"
  },
  "experimental.theme_overrides": {
    "syntax": {
      "operator": {
        "color": "#ebdbb2ff",
        "font_style": null,
        "font_weight": null
      }
    }
  },
  "terminal": {
    "dock": "right"
  },
  "tabs": {
    "git_status": true,
    "file_icons": true
  },
  "relative_line_numbers": true,
  "middle_click_paste": false,
  "languages": {
    "Plain Text": {
      "format_on_save": "off",
      "ensure_final_newline_on_save": false
    },
    "Nix": {
      "tab_size": 2
    }
  },
  "lsp": {
    "starpls": {
      "binary": {
        "path": "starpls",
        "arguments": ["server", "--experimental_infer_ctx_attributes"]
      }
    }
  },

  "tab_size": 4
}
```

**Comment by smitbarmase** [comment]
@Ekleog @cody-quinn @athre0z so the Nix files issue you're seeing happens because we use Shell Script as an injection language in Nix files (kind of like how Markdown can have other languages embedded in it).

Looks like there's a bug in the [Nix Extension](https://github.com/zed-extensions/nix/tree/main) where some code is getting detected as Shell Script when it shouldn't be. This really needs to be fixed in the extension itself, specifically here: [nix/injections](https://github.com/zed-extensions/nix/blob/main/languages/nix/injections.scm). Might be worth opening an issue there for someone who knows Nix and Treesitter queries to tackle. For now, you can work around it by setting `tab_size` for Shell Script instead:
```jsonc
    "Shell Script": {
      "tab_size": 2
    },
```

---

Apart from Nix, I don't see anyone else recently mentioning `tab_size` being broken for other languages. For most folks, switching from `language_overrides` to the `languages` key fixed things up. I'll go ahead and close this issue as resolved.

If you still run into any `tab_size` problems that aren't extension-related, feel free to file a new report. You can tag this issue so I get notified.
